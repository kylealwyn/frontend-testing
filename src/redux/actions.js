import fetch from 'isomorphic-fetch';
import * as Types from './types';

export const addTodo = (text) => ({
  type: Types.ADD_TODO,
  payload: text,
});

function fetchTodosRequest() {
  return {
    type: Types.FETCH_TODOS_REQUEST
  }
}

function fetchTodosSuccess(body) {
  return {
    type: Types.FETCH_TODOS_SUCCESS,
    payload: body
  }
}

function fetchTodosFailure(error) {
  return {
    type: Types.FETCH_TODOS_FAILURE,
    payload: error
  }
}

export function fetchTodos() {
  return dispatch => {
    dispatch(fetchTodosRequest())
    return fetch('http://example.com/todos')
      .then(res => res.json())
      .then(json => dispatch(fetchTodosSuccess(json.body)))
      .catch(ex => dispatch(fetchTodosFailure(ex)))
  }
}
