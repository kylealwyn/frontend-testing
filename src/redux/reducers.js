import { ADD_TODO } from './types'

export const initialTodoState = []

export default function todos(state = initialTodoState, action) {
  switch (action.type) {
    case ADD_TODO:
      return [
        {
          id: state.reduce((maxId, todo) => Math.max(todo.id, maxId), -1) + 1,
          completed: false,
          text: action.text
        },
        ...state
      ]

    default:
      return state
  }
}
