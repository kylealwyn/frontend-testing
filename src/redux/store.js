import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import todoReducer from './reducers';

export const configureStore = () => {
  const store = createStore(todoReducer, {}, applyMiddleware(
   thunkMiddleware,
 ))

 return store;
}
