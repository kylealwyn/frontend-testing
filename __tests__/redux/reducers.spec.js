import todoReducer, { initialTodoState } from '../../src/redux/reducers'
import * as types from '../../src/redux/types'

describe('todos reducer', () => {
  it('should return the initial state', () => {
    expect(
      todoReducer(undefined, {})
    ).toEqual(initialTodoState)
  })

  it('should handle ADD_TODO', () => {
    // create our first todo
    const firstTodoState = todoReducer(initialTodoState, {
      type: types.ADD_TODO,
      text: 'First Todo'
    });

    expect(firstTodoState).toMatchSnapshot();

    // add a second todo
    const secondTodoState = todoReducer(firstTodoState, {
      type: types.ADD_TODO,
      text: 'Second Todo'
    });

    expect(secondTodoState).toMatchSnapshot();
  });
})
