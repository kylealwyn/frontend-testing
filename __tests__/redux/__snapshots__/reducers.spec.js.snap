// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos reducer should handle ADD_TODO 1`] = `
Array [
  Object {
    "completed": false,
    "id": 0,
    "text": "First Todo",
  },
]
`;

exports[`todos reducer should handle ADD_TODO 2`] = `
Array [
  Object {
    "completed": false,
    "id": 1,
    "text": "Second Todo",
  },
  Object {
    "completed": false,
    "id": 0,
    "text": "First Todo",
  },
]
`;
